packages/blade/src/components/Checkbox/_decisions/decisions.md

```diff
@@ -0,0 +1,373 @@
+      </Checkbox>
+    </CheckboxGroup>
+  );
+}
+```
+
+## Component Architecture
+
+- components/Checkbox
+  - useCheckbox.ts
+  - Checkbox.tsx
+  - CheckboxInput.tsx
+  - CheckboxLabel.tsx
+  - CheckboxLabelText.tsx
+  - CheckboxIcon.tsx
+  - useCheckboxGroup.ts
+  - CheckboxGroupContext.tsx
+  - CheckboxGroupFieldset.tsx
+  - CheckboxGroupLegend.tsx
+  - CheckboxGroupHelperText.tsx
+
+We can use a hook based architecture to abstract out state management, accessibility & props to separate hooks like `useCheckbox` & `useCheckboxGroup` which will work cross platform.
+
+**Benefits Of Hooks based architecture:**
+
+- Easier to abstract out the component logic to reusable pieces.
+- Flexible & Works well with our corss-platform needs.
+- We can expose it to consumers which enables them build their own checkbox like component without much hassle.
+
+<img src="./checkbox-api.png" alt="Checkbox API" width="70%">
+
+**useCheckbox:**
+
+useCheckbox will handle the checkbox's toggle state & return props; 
+
+Here's the signature: 
+
+```ts
+const useCheckbox = (props: CheckboxProps) => {
+  // this useControllable state will abstract out the logic for
+  // controlled & uncontrolled state management
+  const [checked, setChecked] = useControllableState()
+  
+  return {
+    checkboxProps: {
+      onMouseDown, 
+      onClick
+    },
+    inputProps: {
+      accessibilityProps: makeAcccessible(...),
+      ...otherProps
+    },
+    labelProps: {
+      ...otherLabelProps
+    }
+  }
+}
+```
+
+**useCheckboxGroup:**
+
+useCheckboxGroup will return label props, group props & handle all the ID generation for the accessible labels, errors & helper texts.
+
+```ts
+const useCheckboxGroup = (props: CheckboxGroupProps) => {
+  return {
+    labelProps: {
+      ...otherProps
+    },
+    groupProps: {
+      isDisabled, 
+      isReadOnly,
+      ...otherGroupProps
+    }
+  }
+}
+```
+
+**CheckboxGroupContext:**
+
+CheckboxGroupContext will pass down the <CheckboxGroup \/> properties like `isDisabled`, `isReadOnly` to the Checkboxes.
+
+## Accessibility
+
+- Checkbox - https://www.w3.org/WAI/ARIA/apg/patterns/checkbox/
+- CheckboxGroup - The group's accessibility is bit tricky to get right, thus I'll implement and refine it and then document it here. 
+
+## Open Questions
+
+1. Currently we only have `helperText` in Checkbox & CheckboxGroup and when we set `isInvalid` it becomes the `negative` variant. The problem is the content of the `helperText` should also change in certain scenarios to convey the exact error. 
+
+With this approach users will have to handle the logic
+
+```tsx
+const [isInvalid, setInvalid] = React.useState(false);
+
+<Checkbox 
+  isInvalid={isInvalid}
+  helperText={
+    isInvalid
+    ? 'Invalid username provided, contains special chars' 
+    : 'Enter valid username'
+  }
+/>
+```
+
+Will it be better to have a `errorText` prop instead which will automatically be switched internally is `isInvalid` is true? 
```

Comment:
I think this should be the case. Whenever `isInvalid` (I prefer `hasError`) is true, it will ignore the helperText & render the errorText instead. We could even enforce consumers to pass an `errorText` whenever they are using `isInvalid` (`hasError`)